### DISCUSSION TITLE
Prisma Integration Testing with Docker as Testing Database

### DISCUSSION THREAD

**Discussion by kwan19961217** [prisma/prisma#19245] [Q&A] [answered]
### Problem
So I have been following the [official document](https://www.prisma.io/docs/guides/testing/integration-testing) to set up the testing database using Docker for the integration test.

And until this part: 
![image](https://github.com/prisma/prisma/assets/59208523/cb0a120c-6ccf-4706-9534-405171e60be7)

By using Docker detached mode, the `prisma migrate deploy` will not wait for the DB server inside Docker container to be ready.

Therefore, the following error will be returned:
```node
> docker-compose up -d

[+] Running 2/2
 ⠿ Network open-api-nest_default       Created                                                                                                                               0.0s
 ⠿ Container integration-tests-prisma  Started                                                                                                                               0.3s
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Datasource "db": MySQL database "tests" at "localhost:3307"

Error: P1017: Server has closed the connection.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! jp-open-api-nest@0.0.1 test: `npm run docker:up  && dotenv -e .env.test -- npx prisma migrate deploy && dotenv -e .env.test -- jest && npm run docker:down`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the jp-open-api-nest@0.0.1 test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

### Current Solution
I am currently replacing the the test script with the following:
```node
"test": "npm run docker:up && sleep 5 && dotenv -e .env.test -- npx prisma migrate deploy && dotenv -e .env.test -- jest && npm run docker:down",`
```
It forces the `prisma migrate deploy` part to wait 5 seconds after the Docker container is up.

I am wondering if there is any suitable solution for the above issue?

**Comment by ChrisPlease** [comment]
I've been using the `wait-for-it` script found in this article: https://www.prisma.io/blog/testing-series-3-aBUyF8nxAn#make-the-script-wait-until-the-database-server-is-ready

**Reply by kwan19961217** [reply, marked as answer]
Thank you for the sharing. Although the blog's method does not work (I will explain below) but has provided me some leads to finally find out a solution.

### Why It Does Not Work

I have read the blog and was happy with the suggested method until I saw the below:
![image](https://github.com/prisma/prisma/assets/59208523/da499b13-9198-4454-b817-19a0ed55f4ef)
Read the two lines start with `wait-for-it.sh`.

The script did not connect to the DB as the url is not even correct. It actually wait 15 seconds for timeout and therefore database will always be ready after that. It basically behave like `sleep` because it will always timeout.

### Solution

Then I found a similar script that specifically deals with database:
[https://github.com/nickjj/wait-until](https://github.com/nickjj/wait-until)

I replaced it with the `wait-for-it.sh` and I can finally move on from this.

**Reply by ChrisPlease** [reply]
FWIW, I get that message too:

```sh
🟡 - Waiting for database to be ready...
wait-for-it.sh: waiting 15 seconds for postgres://postgres
wait-for-it.sh: timeout occurred after waiting 15 seconds for postgres://postgres
🟢 - Database is ready!
```

But my tests still run successfully:

```sh
 Test Files  3 passed (3)
      Tests  10 passed (10)
   Start at  09:28:11
   Duration  7.25s (transform 107ms, setup 274ms, collect 1.47s, tests 4.92s, environment 0ms, prepare 198ms)
   ```
